netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -813,116 +852,239 @@ _net_ebpf_extension_connection_context_initialize(
             // If we fail to delete it, leave it in the list so that we can try again later
             continue;
         }
 
         list_entry = list_entry->Blink;
         RemoveEntryList(&entry->list_entry);
 
-        _net_ebpf_ext_connect_context_count--;
+        --_net_ebpf_ext_connect_context_count;
 
         NET_EBPF_EXT_LOG_MESSAGE_UINT64(
             NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
             NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
-            "_net_ebpf_ext_purge_lru_contexts_under_lock: Delete",
+            "_net_ebpf_ext_purge_connect_contexts_under_lock: Delete",
             entry->transport_endpoint_handle);
 
         ExFreePool(entry);
     }
 
+    // Free entries from low-memory list. These entries should be returned to the free list.
+    list_entry = _net_ebpf_ext_low_memory_blocked_connect_context_list.Blink;
+    while (list_entry != &_net_ebpf_ext_low_memory_blocked_connect_context_list) {
+        net_ebpf_extension_connection_context_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
+        if (!delete_all && entry->timestamp > expiry_time) {
+            break;
+        }
+
+#pragma warning(suppress : 6001) /* entry and list entry are non-null */
+        list_entry = list_entry->Blink;
+        RemoveEntryList(&entry->list_entry);
+
+        if (delete_all) {
+            // Free the memory
+            ExFreePool(entry);
+        } else {
+            // Return the entry to the free list
+            InsertHeadList(&_net_ebpf_ext_low_memory_free_connect_context_list, &entry->list_entry);
+        }
+    }
+
+    // Free pre-allocated entries if we are shutting down.
+    if (delete_all) {
+        while (!IsListEmpty(&_net_ebpf_ext_low_memory_free_connect_context_list)) {
+            PLIST_ENTRY entry = RemoveHeadList(&_net_ebpf_ext_low_memory_free_connect_context_list);
+            net_ebpf_extension_connection_context_t* context =
+                CONTAINING_RECORD(entry, net_ebpf_extension_connection_context_t, list_entry);
+            ExFreePool(context);
+        }
+    }
+
     NET_EBPF_EXT_LOG_MESSAGE_UINT64(
         NET_EBPF_EXT_TRACELOG_LEVEL_INFO,
         NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
-        "_net_ebpf_ext_purge_lru_contexts_under_lock",
+        "_net_ebpf_ext_purge_connect_contexts_under_lock",
         _net_ebpf_ext_connect_context_count);
 }
 
 static void
-_net_ebpf_ext_purge_lru_contexts(BOOLEAN delete_all)
+_net_ebpf_ext_purge_connect_contexts(BOOLEAN delete_all)
 {
     KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
 
-    _net_ebpf_ext_purge_lru_contexts_under_lock(delete_all);
-    ebpf_hash_table_destroy(_net_ebpf_ext_connect_context_hash_table);
-    _net_ebpf_ext_connect_context_hash_table = NULL;
+    _net_ebpf_ext_purge_connect_contexts_under_lock(delete_all);
+    if (delete_all) {
+        ebpf_hash_table_destroy(_net_ebpf_ext_blocked_connect_context_hash_table);
```

Comment:
nit (non-blocking): maybe move this code (deletion of hash table) to a different function and call that function directly from `net_ebpf_ext_sock_addr_unregister_providers`